src/fstab-generator/fstab-generator.c

```diff
@@ -435,7 +436,7 @@ static int add_mount(
 
         /* Order the mount unit we generate relative to the post unit, so that DefaultDependencies= on the
          * target unit won't affect us. */
-        if (post && !FLAGS_SET(flags, AUTOMOUNT) && !FLAGS_SET(flags, NOAUTO))
+        if (post && !FLAGS_SET(flags, MOUNT_AUTOMOUNT) && FLAGS_SET(flags, MOUNT_AUTO))
```

Comment:
So, after some debugging I suspect the issue is right here (`FLAGS_SET(flags, MOUNT_AUTO)`), since we pass `flags = 0` when creating the `sysroot.mount`:

https://github.com/systemd/systemd/blob/7feb1dd6544d1bf373dbe13dd33cf563ed16f891/src/fstab-generator/fstab-generator.c#L737-L746

which with the original expression evaluates to `true` (`!FLAGS_SET(flags, MOUNT_AUTOMOUNT) -> !0 -> 1` and `!FLAGS_SET(flags, NOAUTO) -> !0 -> 1`). However, with `FLAGS_SET(flags, MOUNT_AUTO)` we get `false`, thus skipping the `Before=initrd-root-fs.target` dependency.

`sysroot.mount` **without** this patch:
```
# cat sysroot.mount 
# Automatically generated by systemd-fstab-generator

[Unit]
Documentation=man:fstab(5) man:systemd-fstab-generator(8)
SourcePath=/proc/cmdline
Before=initrd-root-fs.target
Requires=systemd-fsck-root.service
After=systemd-fsck-root.service
After=blockdev@dev-mapper-vg_n47\x2droot.target

[Mount]
Where=/sysroot
What=/dev/mapper/vg_n47-root
Options=ro
```
`sysroot.mount` **with** this patch:
```
# Automatically generated by systemd-fstab-generator

[Unit]
Documentation=man:fstab(5) man:systemd-fstab-generator(8)
SourcePath=/proc/cmdline
Requires=systemd-fsck-root.service
After=systemd-fsck-root.service
After=blockdev@dev-mapper-vg_n47\x2droot.target

[Mount]
Where=/sysroot
What=/dev/mapper/vg_n47-root
Options=ro
```

Reply:
oh, man, you rock. thanks so much for tracking this down. much appreciated!

Reply:
@mrc0mmand nice tracking! :metal: